streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java

```diff
@@ -190,24 +190,72 @@ public void testTimeTracking() {
         record = group.nextRecord(info);
         // 1:[]
         // 2:[6]
-        // st: 4 (doesn't advance because 1 is empty, so it's still reporting the last-known time of 4)
+        // st: 5
         assertEquals(partition1, info.partition());
         verifyTimes(record, 4L, 5L);
         verifyBuffered(1, 0, 1);
         assertEquals(1.0, metrics.metric(lastLatenessValue).metricValue());
 
-        // get one more record, time should not be advanced
+        // get one more record, time should be advanced
         record = group.nextRecord(info);
         // 1:[]
         // 2:[]
-        // st: 4 (1 and 2 are empty, so they are still reporting the last-known times of 4 and 6.)
+        // st: 6
         assertEquals(partition2, info.partition());
         verifyTimes(record, 6L, 6L);
         verifyBuffered(0, 0, 0);
         assertEquals(0.0, metrics.metric(lastLatenessValue).metricValue());
 
     }
 
+    @Test
+    public void testChooseNextRecord() {
+        assertEquals(0, group.numBuffered());
+
+        // add three 3 records with timestamp 1, 5, 3 to partition-1
+        final List<ConsumerRecord<byte[], byte[]>> list1 = Arrays.asList(
+            new ConsumerRecord<>("topic", 1, 1L, recordKey, recordValue),
+            new ConsumerRecord<>("topic", 1, 5L, recordKey, recordValue),
+            new ConsumerRecord<>("topic", 1, 3L, recordKey, recordValue));
+
+        group.addRawRecords(partition1, list1);
+
+        verifyBuffered(3, 3, 0);
+        assertEquals(-1L, group.timestamp());
+        assertEquals(0.0, metrics.metric(lastLatenessValue).metricValue());
+
+        StampedRecord record;
+        final PartitionGroup.RecordInfo info = new PartitionGroup.RecordInfo();
+
+        // get first two records from partition 1
+        record = group.nextRecord(info);
+        assertEquals(record.timestamp, 1L);
+        record = group.nextRecord(info);
+        assertEquals(record.timestamp, 5L);
+
+        // add three 3 records with timestamp 2, 4, 6 to partition-2
+        final List<ConsumerRecord<byte[], byte[]>> list2 = Arrays.asList(
+            new ConsumerRecord<>("topic", 2, 2L, recordKey, recordValue),
+            new ConsumerRecord<>("topic", 2, 4L, recordKey, recordValue),
+            new ConsumerRecord<>("topic", 2, 6L, recordKey, recordValue));
+
+        group.addRawRecords(partition2, list2);
+        // 1:[3]
+        // 2:[2, 4, 6]
+
+        // get one record, next record should be ts=2 from partition 2
+        record = group.nextRecord(info);
+        // 1:[3]
+        // 2:[4, 6]
+        assertEquals(record.timestamp, 2L);
+
+        // get one record, next up should have ts=3 from partition 1 (even though it has larger partition stream time st=5)
```

Comment:
nit: there is not "partition stream time" -- there is "stream time" and "partition time" :)

Reply:
Ack, I chose to use the term to differentiate between the "old" partition time -- largest timestamp observed for a partition so far -- and what is effectively becoming the "new" partition time in this PR, ie timestamp of head record. Will clarify the comment though